# Lesson 3 Assignment
### Jessica Kunkel

In [1]:
#!/usr/bin/env python
""" Import display functions and create a function to print outputs to 
Markdown """

from IPython.display import display, Markdown
def printmd(string):
    display(Markdown(string))

In [2]:
""" Import NumPy, pandas, and dataframe """
import numpy as np
import pandas as pd
from pandas import DataFrame

### 1. Import data mtcars.csv Download mtcars.csvinto Python.

Per the Python file for this week's assignment, I chose to change the working directory of the Jupyter notebook to the directory where I saved the mtcars.csv file. I then used the pandas `read_csv` function to import the mtcars file.

In [3]:
""" Import the os module and set the working directory to where the 
mtcars.csv file is saved """

import os
path = r"C:\Users\jkunk\OneDrive\Documents\_SWENG Masters\9 DAAN 862 - Analytics Programming Python"
os.chdir(path)

In order to better visualize and analyze the data, I chose to set the 'model' attribute as the index column of the dataframe.

In [4]:
""" use read_csv to import mtcars.csv, setting the model attribute 
as the index of the dataframe """

mtcars = pd.read_csv("mtcars.csv", index_col='model')

### 2. Explore the data and perform a statistical analysis of the data
To explore the data, I chose to print the first 5 rows using the `head()` method. I also used the `info()` method to get information on the type of data included in the mtcars dataset. I then chose to perform descriptive statistics on the dataset using the `describe()` method.

In [5]:
""" show the first five rows of the dataframe """

mtcars.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
model                                                                        
Mazda RX4          21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
Mazda RX4 Wag      21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
Datsun 710         22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
Hornet 4 Drive     21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

                   carb  
model                    
Mazda RX4             4  
Mazda RX4 Wag         4  
Datsun 710            1  
Hornet 4 Drive        1  
Hornet Sportabout     2

In [6]:
""" get information on the type of data included in mtcars """

mtcars.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32 entries, Mazda RX4 to Volvo 142E
Data columns (total 11 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   mpg     32 non-null     float64
 1   cyl     32 non-null     int64  
 2   disp    32 non-null     float64
 3   hp      32 non-null     int64  
 4   drat    32 non-null     float64
 5   wt      32 non-null     float64
 6   qsec    32 non-null     float64
 7   vs      32 non-null     int64  
 8   am      32 non-null     int64  
 9   gear    32 non-null     int64  
 10  carb    32 non-null     int64  
dtypes: float64(5), int64(6)
memory usage: 3.0+ KB


In [13]:
""" perform descriptive statistics on the mtcars data """

mtcars.describe()

mpg        cyl        disp          hp       drat         wt  \
count  32.000000  32.000000   32.000000   32.000000  32.000000  32.000000   
mean   20.090625   6.187500  230.721875  146.687500   3.596563   3.217250   
std     6.026948   1.785922  123.938694   68.562868   0.534679   0.978457   
min    10.400000   4.000000   71.100000   52.000000   2.760000   1.513000   
25%    15.425000   4.000000  120.825000   96.500000   3.080000   2.581250   
50%    19.200000   6.000000  196.300000  123.000000   3.695000   3.325000   
75%    22.800000   8.000000  326.000000  180.000000   3.920000   3.610000   
max    33.900000   8.000000  472.000000  335.000000   4.930000   5.424000   

            qsec         vs         am       gear     carb  
count  32.000000  32.000000  32.000000  32.000000  32.0000  
mean   17.848750   0.437500   0.406250   3.687500   2.8125  
std     1.786943   0.504016   0.498991   0.737804   1.6152  
min    14.500000   0.000000   0.000000   3.000000   1.0000  
25%    16.892500   0.000000   0.000000   3.000000   2.0000  
50%    17.710000   0.000000   0.000000   4.000000   2.0000  
75%    18.900000   1.000000   1.000000   4.000000   4.0000  
max    22.900000   1.000000   1.000000   5.000000   8.0000

There are 32 rows in the mtcars dataset. The attributes in the mtcars dataset are mpg (miles per gallon), cyl (number of cylinders), disp (displacement, in cubic inches), hp (gross horsepower), drat (rear axle ratio), wt (weight, in thousands of pounds), qsec (1/4 mile time), vs (engine shape: 0 = v-shaped, 1 = straight), am (transmission: 0 = automatic, 1 = manual), gear (number of forward gears), and carb (number of carburetors).

- the mean mpg is 20.09 +/- 6.027, with a maximum mpg of 33.9 and minimum mpg of 10.4
- the mean number of cylinders is 6.188 +/- 1.786, with a maximum cyl of 8 and minimum cyl of 4
- the mean displacement is 230.7 +/- 123.9 cubic inches, with a maximum disp of 472 and minimum disp of 71.1
- the mean horsepower is 146.7 +/- 68.56, with a maximum hp of 335 and a minimum hp of 52
- the mean rear axle ratio is 3.597 +/- 0.5347, with a maximum drat of 4.93 and a minimum drat of 2.76
- the mean weight is 3.217 +/- 0.9785 thousands of pounds, with a maximum wt of 5.424 and a minimum wt of 1.513
- the mean 1/4 mile time is 17.85 +/- 1.787 seconds, with a maximum qsec of 22.9 and a minimum qsec of 14.5
- the mean, max, and min are less descriptive for the shape of the engine and the transmission. See the following block for summaries of the counts of each type of engine shape and transmission.
- the mean number of forward gears is 3.688 +/- 0.7378, with a maximum gear of 5 and a minimum gear of 3
- the mean number of carburetors is 2.813 +/- 1.615, with a maximum carb of 8 and a minimum carb of 1.

In [15]:
counts_v = mtcars['vs'].value_counts()
counts_t = mtcars['am'].value_counts()

printmd(f"There are {counts_v[0]} cars with v-shaped engines and {counts_v[1]} cars with straight engines in the mtcars dataset.")
printmd(f"There are {counts_t[0]} cars with automatic transmission and {counts_t[1]} cars with manual transmission in the mtcars dataset.")

There are 18 cars with v-shaped engines and 14 cars with straight engines in the mtcars dataset.

There are 19 cars with automatic transmission and 13 cars with manual transmission in the mtcars dataset.

In the mtcars dataset, there are slightly more cars with v-shaped engines than there are with straight engines. There are also slightly more cars with automatic transmission than there are with manual transmission. 

### 3. Analyze mpg for cars with different gear, and show your findings.
To analyze the mpg for cars with different gear, I chose to first group the mtcars dataset by the 'gear' attribute. I was then able to run the descriptive statistics for 'mpg' for each 'gear' group.

In [8]:
""" group the mtcars dataset by 'gear' and select the 'mpg' column,
run descriptive statistics """ 

mpg_gear = mtcars.groupby('gear')['mpg']
mpg_gear.describe()

count       mean       std   min   25%   50%     75%   max
gear                                                            
3      15.0  16.106667  3.371618  10.4  14.5  15.5  18.400  21.5
4      12.0  24.533333  5.276764  17.8  21.0  22.8  28.075  33.9
5       5.0  21.380000  6.658979  15.0  15.8  19.7  26.000  30.4

- There are 15 cars with 3 gears. For these cars, the mean mpg is 16.11 +/- 3.372, with a maximum mpg of 21.5 and a minimum mpg of 10.4.
- There are 12 cars with 4 gears. For these cars, the the mean mpg is 24.53 +/- 5.277, with a maximum mpg of 33.9 and a minimum mpg of 17.8.
- There are 5 cars with 5 gears. For these cars, the the mean mpg is 21.38 +/- 6.659, with a maximum mpg of 30.4 and a minimum mpg of 15.

The data seems to show a positive correlation between the number of gears and the maximum mileage per gallon, but the sample size of the group with 5 gears is smaller, so the higher max mpg value may be an outlier.

### 4. Analyze mpg for cars with different carb, and show your findings.
To analyze the mpg for cars with different carb, I chose to first group the mtcars dataset by the 'carb' attribute. I was then able to run the descriptive statistics for 'mpg' for each 'carb' group.

In [9]:
""" group the mtcars dataset by 'carb' and select the 'mpg' column, 
run descriptive statistics """ 

mpg_carb = mtcars.groupby('carb')['mpg']
mpg_carb.describe()

count       mean       std   min     25%    50%    75%   max
carb                                                              
1       7.0  25.342857  6.001349  18.1  21.450  22.80  29.85  33.9
2      10.0  22.400000  5.472152  15.2  18.825  22.10  25.60  30.4
3       3.0  16.300000  1.053565  15.2  15.800  16.40  16.85  17.3
4      10.0  15.790000  3.911081  10.4  13.550  15.25  18.85  21.0
6       1.0  19.700000       NaN  19.7  19.700  19.70  19.70  19.7
8       1.0  15.000000       NaN  15.0  15.000  15.00  15.00  15.0

- There are 7 cars with 1 carburetor. For these cars, the mean mpg is 25.34 +/- 6.001, with a maximum mpg of 33.9 and a minimum mpg of 18.1.
- There are 10 cars with 2 carburetors. For these cars, the mean mpg is 22.40 +/- 5.472, with a maximum mpg of 30.4 and a minimum mpg of 15.2.
- There are 3 cars with 3 carburetors. For these cars, the mean mpg is 16.30 +/- 1.054, with a maximum mpg of 17.3 and a minimum mpg of 15.2.
- There are 10 cars with 4 carburetors. For these cars, the mean mpg is 15.79 +/- 3.911, with a maximum mpg of 21.0 and a minimum mpg of 10.4.
- There is 1 car with 6 carburetors. The mpg for this car is 19.7.
- There is 1 car with 8 carburetors. The mpg for this car is 15.

Looking at the data, there seems to be a negative correlation between the number of carburetors and the mileage per gallon. However, the standard deviation is consistently large, and there are unequal sample sizes between groups, so it's difficult to draw conclusions on the descriptive statistics at face value.

### 5. Find out which attribute has the most impact on mpg.
After some research, I knew that I wanted to focus on using the built-in pandas functionality in this lession, rather than venture into scikit-learn for formal linear regression. As such, I looked into the "Correlation and Covariance" section of Chapter 5. This highlighted the `corrwith()` function, which allows pairwise correlations between a dataframe's columns and another series or dataframe. In this case, I chose to perform pairwise comparisons by correlation between the 'mpg' attribute and the other attributes in the mtcars dataset.

In [10]:
""" use corrwith to examine pairwise correlations between mpg and the
other columns in the mtcars dataframe """

mtcars.corrwith(mtcars.mpg)

mpg     1.000000
cyl    -0.852162
disp   -0.847551
hp     -0.776168
drat    0.681172
wt     -0.867659
qsec    0.418684
vs      0.664039
am      0.599832
gear    0.480285
carb   -0.550925
dtype: float64

The pairwise correlations show the following:
- a strong negative correlation between the vehicle's weight (wt) and its mpg (-0.867659)
- a strong negative correlation between the vehicle's number of cylinders (cyl) and its mpg (-0.852162)
- a strong negative correlation between the vehicle's displacement (disp) and its mpg (-0.847551)

However, *correlation does not imply causation*, so I decided to delve into creating a multiple linear regression model using 'mpg' as the dependent variable.

In [11]:
""" Import and use scikit-learn to create a multiple linear regression model,
using 'mpg' as the dependent variable. """

import sklearn
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

model = linear_model.LinearRegression()

# split into attributes and dependent variable (mpg)
X = mtcars[['cyl', 'disp', 'hp', 'drat', 'wt', 'qsec', 'vs', 'am', 'gear', 'carb']]
y = mtcars['mpg']

# split into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the linear regression model on the attributes (X) and dependent variable (Y)
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Evaluate model performance
r2 = r2_score(y_test, y_pred)
printmd(f"R-squared: {r2}")

# single out the coefficients of the regression model for examination
coefficients = model.coef_

# create a dataframe of the coefficients and their corresponding attributes
coeff_df = DataFrame(coefficients, columns=['coefficient'], index=['cyl', 'disp', 'hp', 'drat', 'wt', 'qsec', 'vs', 'am', 'gear', 'carb'])
coeff_df

R-squared: 0.7466453084791007

coefficient
cyl     -0.476250
disp     0.018311
hp      -0.009500
drat     1.002130
wt      -4.862037
qsec     1.840952
vs      -1.937186
am       1.076624
gear     1.327108
carb     0.147131

The R-squared value is 0.7466, indicating that the outcome can be predicted with some error from the independent variables.

The attribute with the greatest (positive or negative) coefficient value is weight (wt), indicating that the weight attribute has the greatest influence on mpg.